README.md
<div style="display: block; margin-left: auto;margin-right: auto;width: 40%;">
  <a href="https://plume-oss.github.io/plume-docs/">
    <img src="https://plume-oss.github.io/plume-docs/assets/images/logo-text.png" width="350" alt="Plume Banner">
  </a>
</div>

Plume is a language front-end to construct
an [intermediate representation](https://en.wikipedia.org/wiki/Intermediate_representation) called
a [code-property graphs](https://github.com/ShiftLeftSecurity/codepropertygraph) from JVM bytecode. Plume is graph
database agnostic and can store code-property graphs to multiple graph databases.

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
![GitHub Actions](https://github.com/plume-oss/plume/workflows/CI/badge.svg)
[![codecov](https://codecov.io/gh/plume-oss/plume/branch/master/graph/badge.svg?token=4WY0U6QCU6)](https://codecov.io/gh/plume-oss/plume)
[![](https://jitpack.io/v/plume-oss/plume.svg)](https://jitpack.io/#plume-oss/plume)

## Important

Plume is the original implementation of 
[jimple2cpg](https://github.com/joernio/joern/tree/master/joern-cli/frontends/jimple2cpg). The frontend on Joern project
is optimized around OverflowDB and is much more lightweight. This is project focuses on experimenting with incremental
dataflow analysis and comparing database backend performance.

Versions < 0.6.3 of Plume were Kotlin based but versions from 1.0.0 onwards have been moved to a Scala implementation
for better interfacing with the [CPG schema library](https://github.com/ShiftLeftSecurity/codepropertygraph).

If your project depends on Plume I am happy to still provide maintenance and support but I recommend any new 
research to begin on Joern where I also spend time providing help and support.

## Quickstart

One can run Plume from the `plume` binary which will use `OverflowDB` as the graph database backend if no config is 
found. If one would like to configure another backend then the example `driver.yaml` can be adjusted to include the use 
of another database by uncommenting and editing the respective fields given by the skeleton. E.g. using Neo4j:

```yaml
database: Neo4j
params:
  hostname: localhost
  port: 7687
  username: neo4j
  password: neo4j
  txMax: 25
```

For more documentation and basic guides, check out the [project homepage](https://plume-oss.github.io/plume-docs/) or
the [ScalaDoc](https://plume-oss.github.io/plume/latest/api/io/github/plume/oss/index.html).

*Important*: If you are using the TigerGraph driver you need to install the `gsql_client.jar`and add it to an 
environment variable called GSQL_CLIENT. Instructions are 
[here](https://docs.tigergraph.com/tigergraph-server/current/gsql-shell/using-a-remote-gsql-client) e.g.,

```bash
curl https://docs.tigergraph.com/tigergraph-server/current/gsql-shell/_attachments/gsql_client.jar --output gsql_client.jar
export GSQL_HOME=`pwd`/gsql_client.jar
```

Remember to set the `tgVersion` correctly in the `TigerGraphDriver`.

## Community

* If you have any questions or want to be involved then check out
  our [discussions page](https://github.com/plume-oss/plume/discussions).
* Joern's [Discord](https://discord.gg/28uCANEkK2). Note, this will give you temporary membership
  to the server. Once joined you can obtain permanent membership by being assigned role if necessary. 
* Plume is primarily maintained by [David Baker Effendi](https://davidbakereffendi.github.io/)
    * DM on [Twitter](https://twitter.com/SDBakerEffendi)
    * Email at dbe@sun.ac.za

## Known Bugs

- Due to module encapsulation in Java 17, Kryo serialization for `TinkerGraphDriver` will not work due to serialization
  errors.

## Adding Plume as a Dependency

Replace `X.X.X` with the desired version on [JitPack](https://jitpack.io/#plume-oss/plume).

```sbt
libraryDependencies ++= Seq(
  com.github.plume-oss %% plume % X.X.X
)
```

Don't forget to include the JCenter and JitPack repository in your `build.sbt`.

```sbt
resolvers += "jitpack" at "https://jitpack.io"
```

## Building from Source

Plume releases are available on JitPack. If downloading from JitPack is not an option, or you would like to depend on a
modified version of Plume, you can build Plume locally and use it as an unmanaged dependency. JDK version 11 or higher
is required.

```shell script
git clone https://github.com/plume-oss/plume.git
cd plume
sbt stage
```

This will build `target/scala-2.13/plume_2.13-X.X.X.jar` which can be imported into your local project.

## Logging

Plume uses [SLF4J](http://www.slf4j.org/) as the logging fascade.

## Sponsored by

![Amazon Science](https://assets.amazon.science/dims4/default/ce84994/2147483647/strip/true/crop/1200x630+0+0/resize/1200x630!/quality/90/?url=http%3A%2F%2Famazon-topics-brightspot.s3.amazonaws.com%2Fscience%2F32%2F80%2Fc230480c4f60a534bc077755bae7%2Famazon-science-og-image-squid.png)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.2.7] - 2022-08-29

### Added

- Support for Java source file via Soot's source parser. This can be disabled via 
`Jimple2Cpg.createCpg(parseJavaSource)`.

## [1.2.6] - 2022-05-31

### Changed

- Upgraded `Joern` with latest data flow engine fix and `TypeDecl` modifiers.

## [1.2.5] - 2022-04-25

### Changed

- Upgraded `Joern` with the latest support for annotations and standardized alloc sites.

## [1.2.4] - 2022-04-13

### Changed

- Upgraded `Joern` with the latest bug fixes in JVM bytecode frontend.

## [1.2.3] - 2022-04-05

### Changed

- Added ability to disable cache sharing on `flowsBetween`.
- Simplified `EngineContext` and `Semantics` on `OverflowDbDriver` startup.
- `OverflowDbDriver::methodSemantics` now public.

## [1.2.2] - 2022-04-05

### Changed

- Upgraded Joern version to include configurations that disable cache sharing.

## [1.2.1] - 2022-04-01

### Changed

- `OverflowDbDriver::flowsBetween` performance improvement on initial cache preparation.

## [1.2.0] - 2022-03-31

### Changed

- `OverflowDbDriver` now takes a `DataFlowCacheConfig` argument that specifies 
data flow engine specific configurations.
- `OverflowDbDriver::nodesReachableBy` renamed to `flowsBetween` and now takes functions
as `sources` and `sinks` parameters.

## [1.1.19] - 2022-03-30

### Changed

- Improved data-flow caching performance by holding the same pointer as the initial cache 
and only converting to a serializable form later. 

## [1.1.18] - 2022-03-30

### Added

- New `PlumeStastistics` entries related to measuring result re-use and fetching speeds.

### Changed

- Data flow cache now writes to `.cbor` instead of `.json` for improved I/O performance.

## [1.1.17] - 2022-03-29

### Changed

- `OverflowDbDriver::nodesReachableBy` does not initialize the data-flow engine with the last query's cache if results
are not being re-used.

## [1.1.16] - 2022-03-29

### Added

- Made data flow cache compression optional.

### Changed

- `OverflowDbDriver::nodesReachableBy` now filters paths where first and last node refer to the same immediate parent
node and paths no longer than 1 path element.

## [1.1.15] - 2022-03-28

### Added

- Added statistic to measure the cost of removing outdated graph information.

## [1.1.14] - 2022-03-26

### Changed

- Data flow cache serialization is now buffered and uses LZ4 instead of GZIP.

## [1.1.13] - 2022-03-25

### Changed

- Upgraded Joern version and adjusted to `x2cpg` changes.

## [1.1.12] - 2022-03-24

### Fixed

- Do not write a data-flow cache file if no results were saved.
- Data flow engine context regenerated with latest cache after each query.

## [1.1.11] - 2022-03-24

### Fixed

- Change detection bug where temporary directory where classes were moved to were being accounted for.

## [1.1.10] - 2022-03-23

### Added

- `OverflowDbDriver::nodesReachableBy` now has `sanitizers` parameter for method calls to filter paths out with.

### Changed

- Upgraded Joern and moved to using the `ProgramHandlingUtil` on that `jimple2cpg`.

### Fixed

- Reworked data flow deserialization to address Jackson unable to deserialize `Option[Long]`.
- Fixed major bug where expired node keys were being kept in data flow storage.

## [1.1.9] - 2022-03-22

### Fixed

- `domain::isNodeUnderTypes` not handling potential NPE.

## [1.1.8] - 2022-03-22

### Changed

- Upgraded Joern to 1.1.637 and CPG to 1.3.519.

### Fixed

- `ProgramHandlingUtil::clean` no longer refers to a transitive dependency.

## [1.1.7] - 2022-03-18

### Changed

- Upgraded Joern to 1.1.628 and CPG to 1.3.517.

### Fixed

- Misplaced where `createCpg` would abort on the `sootOnly` config. This has been fixed.

## [1.1.6] - 2022-03-18

### Added

- `OverflowDbDriver::exportAsGraphML` for writing the graph instance as an XML file.

## [1.1.5] - 2022-03-18

### Changed

- Only adding changed files to Soot to improve performance.
- Simplified `PlumeStatistics` file related changes to only those class/methods changed.
- `PlumeDynamicCallLinker` now extends `SimpleCpgPass`.

## [1.1.4] - 2022-03-18

### Added

- Mention that Kryo should only run in Java < 17 in README.

### Changed

- Upgraded Soot to 4.3.0, TinkerGraph/Gremlin to 3.4.11, Joern to 1.1.622.
- Making the reference DB `None` by default to avoid accidentally generating large amounts of stored graphs.

## [1.1.3] - 2022-03-17

### Changed

- Delete the unpacking dir after clean and generating a new one as needed if a successive analysis occurs.
- Making the reference DB `None` by default to avoid accidentally generating large amounts of stored graphs.

## [1.1.2] - 2022-03-16

### Added

- Two new metrics that are tracked: `PROGRAM_CLASSES` and `PROGRAM_METHODS`.

## [1.1.1] - 2022-03-16

### Added

- Two new metrics that are tracked: `CHANGED_CLASSES` and `CHANGED_METHODS`.

## [1.1.0] - 2022-03-15

### Fixed

- Logger classpath issues by removing TigerGraph GSQL client from dependencies and only using SLF4J-API.

### Changed

- Now calling `gsql_client` directly from command line as a process.
- Upgraded TigerGraph version to 3.5.0.

## [1.0.21] - 2022-03-13

### Fixed

- No longer set `deleteOnExit` for files in temp dir since they get cleaned via try-final call already.
- Set Soot to non-application mode and non-whole program mode for efficiency.

## [1.0.20] - 2022-03-11

### Fixed

- `OverflowDB` from throwing `unable to calculate occurrenceCount` exceptions.

## [1.0.19] - 2022-03-11

### Added

- `OverflowDb::safeRemove` to handle exceptions when deleting nodes.

### Changed

- Updated Joern, CPG, and Logback versions.

## [1.0.18] - 2022-03-09

### Changed

- Importing `jimple2cpg` instead of duplicating work here.
- Removed parsing Jimple `IdentityStmt` since they end up duplicating parameters as locals.

src/main/scala/com/github/plume/oss/Plume.scala
package com.github.plume.oss

import better.files.File
import com.github.plume.oss.drivers._
import com.github.plume.oss.util.DataFlowCacheConfig
import io.circe.Json
import io.joern.x2cpg.{X2Cpg, X2CpgConfig}
import scopt.OParser

import java.io.InputStreamReader

/** Command line configuration parameters
  */
final case class Config(
    inputPaths: Set[String] = Set.empty,
    outputPath: String = X2CpgConfig.defaultOutputPath
) extends X2CpgConfig[Config] {

  override def withAdditionalInputPath(inputPath: String): Config =
    copy(inputPaths = inputPaths + inputPath)
  override def withOutputPath(x: String): Config = copy(outputPath = x)
}

/** Entry point for command line CPG creator
  */
object Plume extends App {

  private val frontendSpecificOptions = {
    val builder = OParser.builder[Config]
    import builder.programName
    OParser.sequence(programName("plume"))
  }

  private def parseDriverConfig(): (DriverConfig, IDriver) = {
    import io.circe.generic.auto._
    import io.circe.yaml.parser
    def interpretConfig(rawConf: Json) = {
      rawConf.as[DriverConfig] match {
        case Left(_)     => (null, new OverflowDbDriver())
        case Right(conf) => (conf, createDriver(conf))
      }
    }
    val f = File("driver.yaml")
    if (!f.exists) {
      println("No driver.yaml found, defaulting to OverflowDB driver.")
      return (null, new OverflowDbDriver())
    }
    parser.parse(new InputStreamReader(f.newInputStream)) match {
      case Left(_)        => (null, new OverflowDbDriver())
      case Right(rawConf) => interpretConfig(rawConf)
    }
  }

  private def createDriver(conf: DriverConfig): IDriver = {
    conf match {
      case _ if conf.database == "OverflowDB" =>
        new OverflowDbDriver(
          storageLocation = Option(conf.params.getOrElse("storageLocation", "cpg.odb")),
          heapPercentageThreshold = conf.params.getOrElse("heapPercentageThreshold", "80").toInt,
          serializationStatsEnabled =
            conf.params.getOrElse("serializationStatsEnabled", "false").toBoolean,
          cacheConfig =
            DataFlowCacheConfig(maxCallDepth = conf.params.getOrElse("maxCallDepth", "2").toInt)
        )
      case _ if conf.database == "TinkerGraph" => new TinkerGraphDriver()
      case _ if conf.database == "Neo4j" =>
        new Neo4jDriver(
          hostname = conf.params.getOrElse("hostname", "localhost"),
          port = conf.params.getOrElse("port", "7687").toInt,
          username = conf.params.getOrElse("username", "neo4j"),
          password = conf.params.getOrElse("password", "neo4j"),
          txMax = conf.params.getOrElse("txMax", "25").toInt
        )
      case _ if conf.database == "TigerGraph" =>
        new TigerGraphDriver(
          hostname = conf.params.getOrElse("hostname", "localhost"),
          restPpPort = conf.params.getOrElse("restPpPort", "7687").toInt,
          gsqlPort = conf.params.getOrElse("gsqlPort", "14240").toInt,
          username = conf.params.getOrElse("username", "tigergraph"),
          password = conf.params.getOrElse("password", "tigergraph"),
          timeout = conf.params.getOrElse("timeout", "3000").toInt,
          txMax = conf.params.getOrElse("txMax", "25").toInt,
          scheme = conf.params.getOrElse("scheme", "http")
        )
      case _ if conf.database == "Neptune" =>
        new NeptuneDriver(
          hostname = conf.params.getOrElse("hostname", "localhost"),
          port = conf.params.getOrElse("port", "8182").toInt,
          keyCertChainFile =
            conf.params.getOrElse("keyCertChainFile", "src/main/resources/conf/SFSRootCAC2.pem"),
          txMax = conf.params.getOrElse("txMax", "50").toInt
        )
      case _ =>
        println(
          "No supported database specified by driver.yaml. Supported databases are: OverflowDB, TinkerGraph, Neo4j, Neptune, and TigerGraph."
        ); null
    }
  }

  private val configOption = X2Cpg.parseCommandLine(args, frontendSpecificOptions, Config())
  if (configOption.isEmpty) {
    System.exit(1)
  } else if (configOption.get.inputPaths.size != 1) {
    println("This frontend requires exactly one input path")
    System.exit(1)
  } else {
    val config = configOption.get
    if (config.inputPaths.size == 1) {
      val (conf, driver) = parseDriverConfig()
      if (driver == null) {
        println("Unable to create driver, bailing out...")
        System.exit(1)
      }
      driver match {
        case d: TinkerGraphDriver if conf != null =>
          val importPath = conf.params.get("importPath")
          if (importPath.isDefined) d.importGraph(importPath.get)
        case _ =>
      }
      val cpg = new Jimple2Cpg().createCpg(
        config.inputPaths.head,
        Some(config.outputPath),
        driver
      )
      driver match {
        case d: TinkerGraphDriver if conf != null =>
          val exportPath = conf.params.get("exportPath")
          if (exportPath.isDefined) d.exportGraph(exportPath.get)
        case _ =>
      }
      cpg.close()
    }
  }

}

case class DriverConfig(database: String, params: Map[String, String])
